Verify Patched constraints

balsama opened the pull request:
We've made the mistake a few times of adding a patch to a module without pinning it to a specific release. Then, when the patch gets committed and the maintainer makes a new release, builds fail because the patch we define no longer applies.

This PR adds a composer script that validates all patched dependencies have pinned constraints and runs that script in the .travis.yml script.

balsama committed:
Add patched constraint class

balsama committed:
Merge branch '8.x-2.x' into patched-constraint

balsama committed:
WIP patched constraints

balsama committed:
execute

balsama committed:
Merge branch '8.x-2.x' into patched-constraint

balsama committed:
Optimized return a bit

balsama commented:
Closing until we work out travis's problem.

balsama committed:
Add message back to error output

balsama committed:
run verify-patched-constraints from proper dir

balsama committed:
clean up

balsama commented:
Fixed the command so that it runs within docroot

phenaproxima commented on src/Composer/PatchedConstraint.php:
```diff
@@ -0,0 +1,85 @@
+   */
+  public static function execute(Event $event) {
+    $root_package = $event->getComposer()->getPackage();
+    $patched_dependencies = self::getPatchedDependencyConstraints($root_package);
```
Nitpick: We should generally use static, not self.

phenaproxima commented on src/Composer/PatchedConstraint.php:
```diff
@@ -0,0 +1,85 @@
+
+    /** @var \Composer\Package\Link $package */
+    foreach ($patched_dependencies as $package) {
+      if (self::packageIsUnpinned($package)) {
```
We should favor static over self.

phenaproxima commented on src/Composer/PatchedConstraint.php:
```diff
@@ -0,0 +1,85 @@
+   * @return bool
+   *   True if the constraint appears to be unpinned.
+   */
+  protected static function packageIsUnpinned(\Composer\Package\Link $package) {
```
This method is missing a description in the doc comment. Also, can the type hint be shorted to Link and used at the top of the file?

phenaproxima commented on src/Composer/PatchedConstraint.php:
```diff
@@ -0,0 +1,85 @@
+      return false;
+    }
+    $constraint = $package->getPrettyConstraint();
+    if ((!is_numeric(substr($constraint, 0, 1))) || (strpos($constraint, '|') !== FALSE) || (strpos($constraint, '*') !== FALSE)) {
```
Wow, this is complex. Can we use a regex instead? It would be relatively simple -- all we'd need to do is this: /^[0-9\\.]+(-alpha|beta|rc[0-9]+)?$/

phenaproxima commented on composer.json:
```diff
@@ -49,7 +49,8 @@
         "package": "Acquia\\Lightning\\Composer\\Package::execute",
-        "release-version": "Acquia\\Lightning\\Composer\\ReleaseVersion::execute"
+        "release-version": "Acquia\\Lightning\\Composer\\ReleaseVersion::execute",
+        "verify-patched-constraints": "Acquia\\Lightning\\Composer\\PatchedConstraint::execute"
```
I think this should be a pre-commit hook, but we can bikeshed on that later.

phenaproxima reviewed (changes requested):
This is really nice. I mostly have nits.

balsama committed:
Addressed nits, moved to git hook, and wrote symlink script that runs on install/update

balsama committed:
tweaked class description to be more accurate

balsama commented:
Agree with everything including moving this to a pre-commit hook:
* Moved to git-hooks/pre-commit which will now fail if verify-patched-dependencies returns a non-zero
* Add an additional composer script (symlink) that automatically symlinks .git/hooks/pre-commit to git-hooks/pre-commit 
* Updated composer.json so that composer symlink is run on install and update

balsama committed:
Added a better comment

phenaproxima reviewed (changes requested):
I don't think we need the Symlink thing to be a full-fledged Composer script. We should be able to just do this as the symlink script: ln -s -f git-hooks/pre-commit .git/hooks/pre-commit

balsama commented:
Fair enough. I'd prefer to add it as a phing target so that we're not duplicating the same code in the install and update sections.

balsama committed:
moved symlink to phing script

phenaproxima commented:
This looks great. I think we should merge it.
